File: Desktop/mediflow_pro/backend/app/models/appointment.py
from sqlalchemy import Column, String, Integer, DateTime, ForeignKey, Enum, Float, Text
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.database import Base
import enum
import uuid

def generate_uuid():
    return str(uuid.uuid4())

class AppointmentStatus(enum.Enum):
    PENDING = "pending"
    CONFIRMED = "confirmed"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    CANCELLED = "cancelled"
    NO_SHOW = "no_show"

class Appointment(Base):
    __tablename__ = "appointments"

    id = Column(String, primary_key=True, default=generate_uuid)
    patient_id = Column(String, ForeignKey("users.id"), nullable=False)
    doctor_id = Column(String, ForeignKey("doctors.id"), nullable=False)
    clinic_id = Column(String, ForeignKey("clinics.id"), nullable=True)  # Made nullable
    appointment_date = Column(DateTime, nullable=False)
    time_slot = Column(String, nullable=False)  # e.g., "09:00-09:30"
    department = Column(String, nullable=False)
    reason = Column(Text, nullable=True)
    status = Column(Enum(AppointmentStatus), default=AppointmentStatus.PENDING)
    queue_position = Column(Integer, default=0)
    estimated_wait_time = Column(Integer, default=0)  # in minutes
    patient_latitude = Column(Float, nullable=True)
    patient_longitude = Column(Float, nullable=True)
    notes = Column(Text, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    # Relationships
    patient = relationship("User", back_populates="patient_appointments", foreign_keys=[patient_id])
    doctor = relationship("Doctor", back_populates="appointments")
    clinic = relationship("Clinic", back_populates="appointments")
    queue = relationship("Queue", back_populates="appointment", uselist=False)

File: Desktop/mediflow_pro/lib/models/appointment.dart
// Removed Firebase import

enum AppointmentStatus { pending, confirmed, inProgress, completed, cancelled }

class Appointment {
  final String id;
  final String patientId;
  final String doctorId;
  final String hospitalId;
  final DateTime appointmentDate;
  final String timeSlot;
  final String department;
  final String reason;
  final String? notes;
  final AppointmentStatus status;
  final bool isEmergency;
  final int queuePosition;
  final double? patientLatitude;
  final double? patientLongitude;
  final DateTime? checkInTime;
  final DateTime? startTime;
  final DateTime? endTime;
  final DateTime createdAt;
  final DateTime updatedAt;

  Appointment({
    required this.id,
    required this.patientId,
    required this.doctorId,
    required this.hospitalId,
    required this.appointmentDate,
    required this.timeSlot,
    required this.department,
    required this.reason,
    this.notes,
    required this.status,
    this.isEmergency = false,
    this.queuePosition = 0,
    this.patientLatitude,
    this.patientLongitude,
    this.checkInTime,
    this.startTime,
    this.endTime,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Appointment.fromMap(Map<String, dynamic> map) {
    return Appointment(
      id: map['id'] ?? '',
      patientId: map['patientId'] ?? '',
      doctorId: map['doctorId'] ?? '',
      hospitalId: map['hospitalId'] ?? '',
      appointmentDate: map['appointmentDate'] != null ? DateTime.parse(map['appointmentDate']) : DateTime.now(),
      timeSlot: map['timeSlot'] ?? '',
      department: map['department'] ?? '',
      reason: map['reason'] ?? '',
      notes: map['notes'],
      status: AppointmentStatus.values.firstWhere(
        (e) => e.name == map['status'],
        orElse: () => AppointmentStatus.pending,
      ),
      isEmergency: map['isEmergency'] ?? false,
      queuePosition: map['queuePosition'] ?? 0,
      patientLatitude: map['patientLatitude']?.toDouble(),
      patientLongitude: map['patientLongitude']?.toDouble(),
      checkInTime: map['checkInTime'] != null ? DateTime.parse(map['checkInTime']) : null,
      startTime: map['startTime'] != null ? DateTime.parse(map['startTime']) : null,
      endTime: map['endTime'] != null ? DateTime.parse(map['endTime']) : null,
      createdAt: map['created_at'] != null ? DateTime.parse(map['created_at']) : DateTime.now(),
      updatedAt: map['updated_at'] != null ? DateTime.parse(map['updated_at']) : DateTime.now(),
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'patientId': patientId,
      'doctorId': doctorId,
      'hospitalId': hospitalId,
      'appointmentDate': appointmentDate.toIso8601String(),
      'timeSlot': timeSlot,
      'department': department,
      'reason': reason,
      'notes': notes,
      'status': status.name,
      'isEmergency': isEmergency,
      'queuePosition': queuePosition,
      'patientLatitude': patientLatitude,
      'patientLongitude': patientLongitude,
      'checkInTime': checkInTime?.toIso8601String(),
      'startTime': startTime?.toIso8601String(),
      'endTime': endTime?.toIso8601String(),
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  Appointment copyWith({
    String? id,
    String? patientId,
    String? doctorId,
    String? hospitalId,
    DateTime? appointmentDate,
    String? timeSlot,
    String? department,
    String? reason,
    String? notes,
    AppointmentStatus? status,
    bool? isEmergency,
    int? queuePosition,
    double? patientLatitude,
    double? patientLongitude,
    DateTime? checkInTime,
    DateTime? startTime,
    DateTime? endTime,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Appointment(
      id: id ?? this.id,
      patientId: patientId ?? this.patientId,
      doctorId: doctorId ?? this.doctorId,
      hospitalId: hospitalId ?? this.hospitalId,
      appointmentDate: appointmentDate ?? this.appointmentDate,
      timeSlot: timeSlot ?? this.timeSlot,
      department: department ?? this.department,
      reason: reason ?? this.reason,
      notes: notes ?? this.notes,
      status: status ?? this.status,
      isEmergency: isEmergency ?? this.isEmergency,
      queuePosition: queuePosition ?? this.queuePosition,
      patientLatitude: patientLatitude ?? this.patientLatitude,
      patientLongitude: patientLongitude ?? this.patientLongitude,
      checkInTime: checkInTime ?? this.checkInTime,
      startTime: startTime ?? this.startTime,
      endTime: endTime ?? this.endTime,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}